### ISSUE TITLE
Error: connection timed out

### ISSUE THREAD

**Issue by raulfunkie** [apexad/homebridge-eufy-robovac#8]
Getting these errors from time to time and it renders my whole HomeKit operation unusable until I restart Homebridge.

Any ideas what could be causing this error and any possible solution?

```Error: connection timed out
    at Socket.<anonymous> (/usr/lib/node_modules/homebridge-eufy-robovac/dist/index.js:1:23097)
    at Object.onceWrapper (events.js:312:28)
    at Socket.emit (events.js:223:5)
    at Socket._onTimeout (net.js:474:8)
    at listOnTimeout (internal/timers.js:531:17)
    at processTimers (internal/timers.js:475:7)```

**Comment by raulfunkie** [comment]
[homebridge-error.log](https://github.com/joshstrange/homebridge-eufy-robovac/files/4231605/homebridge-error.log)
attaching complete homebridge log just in case someone wants to take a look at it for further review.

@aperkins1310 hey, this is the error I was telling you about the other day.

**Comment by MonsterEnergy-wtf** [comment]
Nice to see that someone else is facing the same issue. 
Not sure since what time I have this issue but it seems that the package has some issues with the latest version of homebrige. 
Maybe someone who has more experience with this package could take a look into this here please?
I'm glad to help where ever I can.

**Comment by MonsterEnergy-wtf** [comment]
I just found a "solution". Some other guy did a great job and fixed some issues and uploaded a transpiled verison of the index.js.
https://github.com/joshstrange/homebridge-eufy-robovac/issues/4#issuecomment-538154698
I already asked him/her to share the source so everone can be happy.

Not sure what caused my issue but I guess that homebridge does a dynamic recompile once an updaate occures or something like that.

So basically put the posted index.js to your dist folder and the connection issues are gone.

**Comment by raulfunkie** [comment]
That's actually a change I have in place and I keep having the connection issues. Not sure why it keeps happening. :(

**Comment by MonsterEnergy-wtf** [comment]
I guess it simple to explain (if this solves your issue).
The npm install will be executed once an update of homebridge happened or something like that. In this case the whole dist folder will be wiped and a new ("defect") version of the index.js is created from sources. 
This will then cause the issue.

When you now overwrite the already applied fix to the index.js everything will be fine.
In my case the index.js which was working fine was overwritten by the npm install and then the issues began again.

If the uploader for the fix is able to provide the solution he/she made it would prevent the issues from happening again. Then a new version of this module could also cover our issue and everyone would be happy.

**Comment by MonsterEnergy-wtf** [comment]
Funny. I just realized that there might be a quick fix for this. Maybe someone can confirm / decline this?

What i did:

1. Removed all dist outputs in eufy-robovac and homekit-eufy-robovac packges
2. npm run build -> Gives a list of issues related to outdated packages (at least at homekit-eufy-robovac)
3. npm audit fix (I normally don't do this because it might fu** up your whole package - but what do I have to lose?)
4. removed node_modules again and npm install
5. npm run build. 
6. After this (and an adjustment on the webpack.config.js on **eufy-robovac** and **homekit-eufy-robovac** I was able to build both packages from scratch.
7. Restart homebridge -> For now it doesn't give time outs. 

I have to figure out how to provide the code (webpack.config.js and package.lock and package.json)to a new pull request. I will do this once I can confirm that my stuff is working. 

Can someone give it a shot?

**Comment by apexad** [comment]
What is the adjustments to webpack.config.js you made?  I have done 2 PRs with `npm audit fix` for both packages.

**Comment by apexad** [comment]
@MonsterEnergy-wtf I assume mode?  Let me know if it is not.

**Comment by apexad** [comment]
@MonsterEnergy-wtf I assume you are using `deviceId` and not `ip addresses`.  Seems that either my devices IP address has changed, or IP address no longer works.  However, when I switched to deviceId it worked!

**Comment by apexad** [comment]
Duplicate of #3 #1 #4
This should now be fixed as well. However, please note, it seems IP address is no longer supported by the Tuya API (used by this plugin). Please switch to deviceId if you are using IP address. I will update the plugin README and such to reflect this soon.

**Comment by MonsterEnergy-wtf** [comment]
> What is the adjustments to webpack.config.js you made? I have done 2 PRs with `npm audit fix` for both packages.

I had to change the rules from webpack.config.js on both projects (homebridge-eufy-robovac and eufy-robovac):

`rules: [
	{
		test: /\.tsx?$/,
		use: 'ts-loader',
		exclude: /node_modules/
	}
]`

to

`rules: [
	{
    	test: /\.(ts|tsx)?$/,
        include: path.resolve(__dirname, 'src'),
        use: [
        {
        	loader: 'ts-loader'
        }
        ]
    }
]`

After that I was able to compile it on my system. Not sure what causes the issue. Maybe this is not required to be changed.

**Comment by MonsterEnergy-wtf** [comment]
> @MonsterEnergy-wtf I assume you are using `deviceId` and not `ip addresses`. Seems that either my devices IP address has changed, or IP address no longer works. However, when I switched to deviceId it worked!

Yes. I never use IPs because they can change over time. I add device id and key to my config.

**Comment by apexad** [comment]
@MonsterEnergy-wtf weird, I think needing to fix this is on raspberry-pi, but not on macOS.  Anyway, the plugin should actually be built prior to publish, so this should not be needed as long as `npm run build` works for the developer.

By the way @joshstrange it looks like you were forgetting to do `nom run build` so you probably saw I just pushed out 1.3.1 and 1.1.1 which adds a `prePublish` script so I never forget to do the same thing in the future (and both of these versions have the dist folder published).

@MonsterEnergy-wtf you want to install the official `homebridge-eufy-robovac@1.1.1` from 7pm and confirm everything still works for you?

If it does this is truly resolved and I will work to remove the IP address option and add a `config.schema.json` later today. Thanks for your help in getting this working again!

**Comment by MonsterEnergy-wtf** [comment]
@apexad related to version 1.1.1 I had issues getting the server back running. I updated the package without any issues (this is what the UI tells me). For some reason the old dist folder remainswith old index.js. After a complete restart of the server the server could not restart homebridge cause of missing homebrigde-eufy-robovac/dist/index.js  (whaaaaat?!)

I had to add my changes to the webpack.config.js on homebridge-eufy-robovac/node_modules/eufy-robovac/webpack.config.js and rebuild eufy-robovac.

After that I did the same changes to the homebridge-eufy-robovac/webpack.config.js and voila: everything is working fine.

So making it short: for Ubuntu server and Linux in general we should provide the fix I prepared or something like this. After that people could rebuild the stuff by themselves if they need to. Without the fix they are screwed. 

As far as I can tell you with deviceID and Key it seems to be fine. But I will test it by tomorrow because my family is sleeping and I don't want to disturb them with that damn "Beeeeep". 
I will provide feedback by tomorrow.

**Comment by apexad** [comment]
I wonder if this is a permissions issue with the dist folder since you are running npm run build yourself.

If you can, try completely uninstalling it and confirming the folder is totally gone, then reinstall it.

**Comment by raulfunkie** [comment]
Testing the latest changes (1.1.1). Getting this while running `homebridge -D`:
```
[5/23/2020, 7:43:40 PM] Loaded plugin: homebridge-eufy-robovac@1.1.1
[5/23/2020, 7:43:40 PM] Registering accessory 'homebridge-eufy-robovac.Eufy RoboVac'
[5/23/2020, 7:43:41 PM] [Robotina] Initializing Eufy RoboVac accessory...
Looking for device...
Found device 60608308ecfabc982fba at 192.168.7.248
Connecting...
Connected!
{}
Disconnected!
```
```
[5/23/2020, 7:45:44 PM] [Robotina] getCleanState
Connecting...
Connecting...
Connecting...
Connected!
Connected!
{}
Disconnected!
Disconnected!
```

```
[5/23/2020, 7:46:01 PM] [Robotina] getCleanState
Connecting...
Connecting...
Connecting...
Connected!
{}
Disconnected!
Disconnected!
{}
Disconnected!
Connected!
```

Seems like it connects then disconnects instantly. I was testing some other changes (running `homebridge -D` and it got to report the battery level and whatnot. But I can't get it working anymore. Any ideas?

**Comment by apexad** [comment]
It appears IP address no I longer works. You need to go get your devId/deviceId.

**Comment by raulfunkie** [comment]
I'm actually using deviceID and not IP address for my config.

**Comment by apexad** [comment]
Hmmm, using latest app on your phone and your robot is updated?

**Comment by raulfunkie** [comment]
Yeah, I just "restarted" homebridge (using pm2 now since I finished debugging other things) and it seems to be running well. I've checked the eufy Home app and no updates are pending.

![image](https://user-images.githubusercontent.com/15929687/82742637-b6412300-9d2e-11ea-8f37-5eb2a5b911bf.png)

**Comment by apexad** [comment]
Glad to hear it is working now.

**Comment by raulfunkie** [comment]
@apexad are you planning on getting battery level and charging status implemented again? I saw (when I ran `homebridge -D` before) that the vacuum is reporting battery status (it showed the level in simple numbers (16, 25, etc).

I can offer my device and time for testing, since I don't really know any useful programming language that applies to this.

Thanks for getting it working again!

**Comment by MonsterEnergy-wtf** [comment]
@raulfunkie  The status is already there. In Homekit app you will have to switch to seperate view "show as seperate tiles". Then you will have two tiles per robovac. On the one you will see charging status.

**Comment by apexad** [comment]
> @apexad are you planning on getting battery level and charging status implemented again? I saw (when I ran `homebridge -D` before) that the vacuum is reporting battery status (it showed the level in simple numbers (16, 25, etc).
> 
> I can offer my device and time for testing, since I don't really know any useful programming language that applies to this.
> 
> Thanks for getting it working again!

As far as I can tell, if you look at the fan accessory exposed, Charging and Battery % are working for me. Is it not for you?

**Comment by apexad** [comment]
> @raulfunkie The status is already there. In Homekit app you will have to switch to seperate view "show as seperate tiles". Then you will have two tiles per robovac. On the one you will see charging status.

Yes, this.  I highly recommend the 'Show as Separate Tiles' option.  Do not let HomeKit group them. As mentioned in https://github.com/joshstrange/homebridge-eufy-robovac/issues/2 this seems to cause some weird issues.

**Comment by MonsterEnergy-wtf** [comment]
@apexad Glad we could solved the issues that easy. If you need my help for this project in future please let me know.

**Comment by MonsterEnergy-wtf** [comment]
@apexad What I now can confirm: all of my three Robovacs 30C are working fine.
The only thing I realized: when you miss the `"ip": "",` from the config, all Robovacs will become unresponsive. Not sure what happens here, but it might be cool if we could add this to the template - even if we are not using it. 
An other thing might be to remove it totally from the code and the config.

But overall: I'm happy that we have a working package right now!
Thanks everyone!

**Comment by apexad** [comment]
> @apexad What I now can confirm: all of my three Robovacs 30C are working fine.
> 
> The only thing I realized: when you miss the `"ip": "",` from the config, all Robovacs will become unresponsive. Not sure what happens here, but it might be cool if we could add this to the template - even if we are not using it. 
> 
> An other thing might be to remove it totally from the code and the config.
> 
> 
> 
> But overall: I'm happy that we have a working package right now!
> 
> Thanks everyone!

Weird, I did not need to add this, but yes I will be taking it out from the config.
